workflow/sync/sync_manager.go

```diff
@@ -66,6 +66,77 @@ func (cm *Manager) CheckWorkflowExistence() {
 	}
 }
 
+func getUpgradedKey(wf *wfv1.Workflow, key string, level SyncLevelType) string {
+	if wfv1.CheckHolderKeyVersion(key) == wfv1.HoldingNameV1 {
+		if level == WorkflowLevel {
+			return getHolderKey(wf, "")
+		}
+		return getHolderKey(wf, key)
+	}
+	return key
+}
+
+type SyncLevelType int
+
+const (
+	WorkflowLevel SyncLevelType = 1
+	TemplateLevel SyncLevelType = 2
+	ErrorLevel    SyncLevelType = 3
+)
+
+// legacy keys can be of the form
+// x where x is a workflow name
+// unfortunately this doesn't differentiate between workflow level keys
+// and template level keys. So upgrading is a bit tricky here.
+
+// given a legacy holding name x, namespace y and workflow name z.
+// in the case of a workflow level
+// if x != z
+// upgradedKey := y/z
+// elseif x == z
+// upgradedKey := y/z
+// in the case of a template level
+// if x != z
+// upgradedKey := y/z/x
+// elif x == z
+// upgradedKey := y/z/x
+
+// there is a possibility that
+// a synchronization exists both at the template level
+// and at the workflow level -> impossible to upgrade correctly
+// due to ambiguity. Currently we just assume workflow level.
+func getWorkflowSyncLevelByName(wf *wfv1.Workflow, lockName string) (SyncLevelType, error) {
+	if wf.Spec.Synchronization != nil {
+		syncLockName, err := GetLockName(wf.Spec.Synchronization, wf.Namespace)
+		if err != nil {
+			return ErrorLevel, err
+		}
+		checkName := syncLockName.EncodeName()
+		if lockName == checkName {
+			return WorkflowLevel, nil
+		}
+	}
+
+	var firstErr error
+	for _, template := range wf.Spec.Templates {
+		if template.Synchronization != nil {
+			syncLockName, err := GetLockName(template.Synchronization, wf.Namespace)
+			if err != nil {
+				firstErr = err
+				continue
+			}
+			checkName := syncLockName.EncodeName()
+			if lockName == checkName {
+				return TemplateLevel, nil
+			}
+		}
+	}
+	if firstErr == nil {
```

Comment:
Why store firstErr and then not use it in the error message?

Could be a `bool` instead, but prefer to spit out the error text here.

Reply:
Sure will do.